### PR TITLE
AWS EC2Autoscaler -- Accept Tags

### PR DESCRIPTION
**Pull request by nicktylah**
- Update dynamic worker config to contain "tags" key, which will tag newly provisioned instances on autoscaling events.
- Add tests for legacy configs (with null tags)
- This branch's parent is `feature-ec2autoscaler-region` ( #2550 ) -- please see the second commit for the relevant changes made in this PR. Thanks!

### CODE REVIEW

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java [hunk 1]
```diff
@@ -161,6 +175,27 @@ public String apply(Instance input)
 
       log.info("Created instances: %s", instanceIds);
 
+      if (envConfig.getNodeData().getTags() != null) {
+        final List<Tag> tags = envConfig
+            .getNodeData()
+            .getTags()
+            .entrySet()
+            .stream()
```
Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java]: We're still on java7. See https://groups.google.com/d/msg/druid-development/6I9uhO3ZIeQ/WG-K6JdEEQAJ

Comment (drcrallen) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java]: You should be able to accomplish this with guava stuff though

**Comment by fjy** [comment]
@nicktylah this doesnt seem to compile

**Comment by nicktylah** [comment]
@fjy The compiler problem stems from a dependency I introduced in #2550 (the `server` module depends on the `indexing-service` module for access to necessary configuration settings). Do you have a recommendation for how to access those settings without introducing that dependency?

File: indexing-service/src/test/java/io/druid/indexing/overlord/autoscaling/EC2AutoScalerSerdeTest.java [hunk 2]
```diff
@@ -82,6 +89,12 @@ public Object findInjectableValue(
         EC2AutoScaler.class
     );
     verifyAutoScaler(roundTripAutoScaler);
+    Map<String, String> expectedTags = new HashMap<>();
+    expectedTags.put("name", "joffrey");
```
Comment (xvrl) [inline indexing-service/src/test/java/io/druid/indexing/overlord/autoscaling/EC2AutoScalerSerdeTest.java]: why not put those lines as part of `verifyAutoScaler`?

File: indexing-service/src/test/java/io/druid/indexing/overlord/autoscaling/EC2AutoScalerTest.java [hunk 3]
```diff
@@ -47,17 +49,34 @@
...
+import java.util.Map;
 
 /**
  */
 public class EC2AutoScalerTest
 {
   private static final String AMI_ID = "dummy";
   private static final String INSTANCE_ID = "theInstance";
+  private static final Map<String, String> TAGS = new HashMap<>();
+  static
```
Comment (xvrl) [inline indexing-service/src/test/java/io/druid/indexing/overlord/autoscaling/EC2AutoScalerTest.java]: `ImmutableMap.of(k, v)` would be cleaner than the static block

File: indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java [hunk 4]
```diff
@@ -161,6 +165,24 @@ public String apply(Instance input)
 
       log.info("Created instances: %s", instanceIds);
 
+      if (envConfig.getNodeData().getTags() != null) {
+        final List<Tag> tags = Lists.transform(
+            new ArrayList<>(envConfig.getNodeData().getTags().entrySet()),
```
Comment (xvrl) [inline indexing-service/src/main/java/io/druid/indexing/overlord/autoscaling/ec2/EC2AutoScaler.java]: no need to create a new list, you can directly apply `Collections2.transform` on the entrySet and use the resulting collection on the `withTags` call.

File: server/src/main/java/io/druid/indexing/overlord/autoscaling/ResourceManagementSchedulerConfig.java [hunk 5]
```diff
@@ -28,6 +28,9 @@
 public class ResourceManagementSchedulerConfig
 {
   @JsonProperty
+  private String region = "us-east-1";
```
Comment (xvrl) [inline server/src/main/java/io/druid/indexing/overlord/autoscaling/ResourceManagementSchedulerConfig.java]: I believe region should be configured as part of `EC2EnvironmentConfig`. The `ResourceManagementScheduler` is agnostic of the underlying autoscaler and should not depend anything AWS specific.

File: server/src/main/java/io/druid/guice/AWSModule.java [hunk 6]
```diff
@@ -47,8 +50,17 @@ public AWSCredentialsProvider getAWSCredentialsProvider(final AWSCredentialsConf
...
+  public AmazonEC2 getEc2Client(
+      AWSCredentialsProvider credentials,
+      ResourceManagementSchedulerConfig resourceManagementSchedulerConfig)
   {
-    return new AmazonEC2Client(credentials);
+    final AmazonEC2Client amazonEC2Client = new AmazonEC2Client(credentials);
+    final String region = resourceManagementSchedulerConfig.getRegion();
+    if (region != null) {
+      try {
+        amazonEC2Client.setRegion(Region.getRegion(Regions.fromName(region)));
```
Comment (xvrl) [inline server/src/main/java/io/druid/guice/AWSModule.java]: the problem with changing region globally is that this will also change the region for other AWS services (such as S3), not just autoscaling, and could cause some unintended behavior.
I believe we should have separate ec2 client providers for auto-scaling and s3, and be able to configure region separately.
